### PR TITLE
WIP: [feature] Add support for hardware ID / serial number

### LINKED ISSUE #86: [feature] Hardware id field
Follow up to mailing list discussion: https://groups.google.com/d/msg/openwisp/CGJKYfIregU/yd3-ma9GAAAJ

We should add a dedicated field in django-netjsonconfig with a generic name which can be used for any type of hardware ID. Since it's generic, we should be flexible in defining this field, but we know for sure it's going to be a unique alphanumeric string with a generous length, the database field can be NULL (and if an empty string is supplied it should be converted to None to avoid triggering the unique DB constraint error).

The field may or may not be required depending on a configuration (django setting), which makes it not required by default, which maintains backward compatibility.

The human readable label and the help text of the field should be configurable, by default we could use something like "Serial number", but it can be modified to use a different term depending on the context where it is used.

In my opinion this feature should be deactivated by by default and shown only when a specific setting is turned on.

When the feature is enabled, the following things should happen:

- it should be possible to search devices in the admin by their serial number
- it should be shown in the device list, probably as first field from the left
- it should be shown and be editable in the add device page, shown right before or right after the name field 
- it should be shown and (probably) readonly in change device page, shown right before or right after the name field
- the controller view which allows registration of new devices should save this value if it finds it in the HTTP POST data (this may already work now due to how that part is implemented)
- the field should become available in the device context

### PR DESCRIPTION
**Pull request by okraits**
Closes #86

### CODE REVIEW

File: django_netjsonconfig/controller/generics.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def init_object(self, **kwargs):
         config_model = device_model.get_config_model()
         options = {}
         for attr in kwargs.keys():
+            print("###", attr)
```
Comment (nemesisdesign) [inline django_netjsonconfig/controller/generics.py]: forgot this here?

Comment (okraits) [reply django_netjsonconfig/controller/generics.py]: I know this - I'm not done yet ;-)

**Comment by nemesisdesign** [comment]
PS: rebase also on the current master when you can

**Comment by okraits** [comment]
@nemesisdesign I fixed the README.rst and rebased on the current master.

For the `__str__` method, do you mean overwriting it in the class https://github.com/openwisp/django-netjsonconfig/blob/master/django_netjsonconfig/models.py#L16? I didn't find a `__str__` method where the mac address is used. Is this done automatically?

**Comment by nemesisdesign** [comment]
@okraits you're right is not the macaddress that it's used but the name, so if in openwisp-config you're sending the serial number as the default name you'll get the result you want

**Comment by okraits** [comment]
@nemesisdesign hardware ID/serial number and name are separate fields in openwisp-config:
https://github.com/openwisp/openwisp-config/pull/75/commits/3993bc523094a8876c7c0fe7b1e7398c78bf7224#diff-b00d48fc9d162653819b99c7ae0a9f5eR167

I added a test like you said but it fails badly. I don't understand the test. Could you please help me with it?

**Pull request merged**